File: pkm_rs/src/pkm/plugins/cfru/pk3cfru.rs
use super::conversion::util::{from_gen3_cfru_pokemon_index, to_gen3_cfru_pokemon_index};
use super::conversion::{NATIONAL_DEX_TO_RADICAL_RED_MAP, RADICAL_RED_TO_NATIONAL_DEX_MAP};
use crate::pkm::traits::IsShiny;
use crate::pkm::{Error, Pkm, Result};
use crate::resources::{
    Ball, FormeMetadata, GameOfOriginIndex, MoveSlot, SpeciesAndForme, SpeciesMetadata,
};
use crate::strings::Gen3String;
use crate::substructures::{Gender, MarkingsFourShapes, Stats8};
use crate::util;

use serde::Serialize;

pub const CFRU_BALLS: [Ball; 27] = [
    Ball::Master,
    Ball::Ultra,
    Ball::Great,
    Ball::Poke,
    Ball::Safari,
    Ball::Net,
    Ball::Dive,
    Ball::Nest,
    Ball::Repeat,
    Ball::Timer,
    Ball::Luxury,
    Ball::Premier,
    Ball::Dusk,
    Ball::Heal,
    Ball::Quick,
    Ball::Cherish,
    Ball::None, // INVALID in TS, mapped to None here
    Ball::Fast,
    Ball::Level,
    Ball::Lure,
    Ball::Heavy,
    Ball::Love,
    Ball::Friend,
    Ball::Moon,
    Ball::PokeLegendsArceus, // PokeHisui in TS
    Ball::Beast,
    Ball::Dream,
];

#[inline]
fn cfru_ball_from_index(idx: u8) -> Ball {
    CFRU_BALLS.get(idx as usize).copied().unwrap_or(Ball::Poke)
}

#[inline]
fn cfru_ball_index(ball: Ball) -> u8 {
    if let Some(i) = CFRU_BALLS.iter().position(|&b| b == ball) {
        i as u8
    } else if (Ball::PokeLegendsArceus as u8) <= (ball as u8)
        && (ball as u8) <= (Ball::Origin as u8)
    {
        // Collapse all Hisui/LA balls to PokeLegendsArceus for CFRU writes (TS fallback).
        CFRU_BALLS
            .iter()
            .position(|&b| b == Ball::PokeLegendsArceus)
            .unwrap() as u8
    } else {
        // Fallback to Poke on unknown.
        CFRU_BALLS.iter().position(|&b| b == Ball::Poke).unwrap() as u8
    }
}

/// PK3CFRU (58 bytes)
#[derive(Debug, Default, Serialize, Clone)]
pub struct Pk3cfru {
    // Personality 0:4
    pub personality_value: u32,

    // OTID 4:8
    pub trainer_id: u16, // 0..6
    pub secret_id: u16,  // 6..8

    // Nickname 8:18
    pub nickname: Gen3String<10>,

    // Language 18
    pub language_index: u8,

    // Sanity 19
    // pub sanity: u8,

    // OT Name 20:27
    pub trainer_name: Gen3String<7>,

    // Markings 27
    pub markings: MarkingsFourShapes,

    // Species 28:30
    pub species_and_forme: SpeciesAndForme,
    cfru_species_index: u16, // raw CFRU game index

    // Held Item 30:32
    pub held_item_index: u16,

    // Exp 32:36
    pub exp: u32,

    // Move PP Up 36 (2 bits each)
    pub move_pp_ups: [u8; 4],

    // Friendship 37
    pub trainer_friendship: u8,

    // Pokeball 38
    pub ball: Ball,

    // Moves 38:43 (5 bytes total for 4 moves with 10 bits each)
    pub moves: [MoveSlot; 4],
    pub move_pp: [u8; 4], // computed; not stored in CFRU bytes

    // EVs 43:49
    pub evs: Stats8,

    // 49
    pub pokerus_byte: u8,

    // 50
    pub met_location_index: u8,

    // 51:53
    pub met_level: u8,                     // 0x34 bits 0..6
    pub game_of_origin: GameOfOriginIndex, // 0x34 bits 7..10
    pub can_gigantamax: bool,              // 0x34 bit 11
    pub trainer_gender: Gender,            // 0x34 bit 15

    // 53:57
    pub ivs: Stats8,              // 0x36..0x39 (30 bits)
    pub is_egg: bool,             // 0x36 bit 30
    pub has_hidden_ability: bool, // 0x36 bit 31

    // Values not stored in CFRU bytes
    pub is_nicknamed: bool,
    pub current_hp: u16,
}

impl Pk3cfru {
    pub const BOX_SIZE: usize = 58;
    pub const PARTY_SIZE: usize = 100;

    /// Unpack 4×10-bit move IDs from 5 bytes at 0x27..0x2B (little-endian bits).
    #[inline]
    fn read_moves_10bit(bytes: &[u8]) -> [MoveSlot; 4] {
        let base = 0x27;
        let v = (bytes[base] as u64)
            | ((bytes[base + 1] as u64) << 8)
            | ((bytes[base + 2] as u64) << 16)
            | ((bytes[base + 3] as u64) << 24)
            | ((bytes[base + 4] as u64) << 32);

        [
            MoveSlot::from((v & 0x3FF) as u16),
            MoveSlot::from(((v >> 10) & 0x3FF) as u16),
            MoveSlot::from(((v >> 20) & 0x3FF) as u16),
            MoveSlot::from(((v >> 30) & 0x3FF) as u16),
        ]
    }

    /// Pack 4×10-bit move IDs into 5 bytes at 0x27..0x2B (little-endian bits).
    #[inline]
    fn write_moves_10bit(moves: &[MoveSlot; 4], bytes: &mut [u8]) {
        let base = 0x27;
        let m0 = u16::from(moves[0]) as u64 & 0x3FF;
        let m1 = u16::from(moves[1]) as u64 & 0x3FF;
        let m2 = u16::from(moves[2]) as u64 & 0x3FF;
        let m3 = u16::from(moves[3]) as u64 & 0x3FF;

        let v = (m0) | (m1 << 10) | (m2 << 20) | (m3 << 30);

        bytes[base] = (v & 0xFF) as u8;
        bytes[base + 1] = ((v >> 8) & 0xFF) as u8;
        bytes[base + 2] = ((v >> 16) & 0xFF) as u8;
        bytes[base + 3] = ((v >> 24) & 0xFF) as u8;
        bytes[base + 4] = ((v >> 32) & 0xFF) as u8;
    }

    pub fn from_bytes(bytes: &[u8]) -> Result<Self> {
        let size = bytes.len();
        if size < Self::BOX_SIZE {
            return Err(Error::ByteLength {
                expected: Self::BOX_SIZE,
                received: size,
            });
        }

        // 0x1C..0x1E: CFRU game species index
        let cfru_species_index = u16::from_le_bytes(bytes[0x1C..0x1E].try_into().unwrap());
        let saf =
            from_gen3_cfru_pokemon_index(cfru_species_index, &RADICAL_RED_TO_NATIONAL_DEX_MAP)?;

        // 0x34..0x36: met info & flags
        let meta = u16::from_le_bytes(bytes[0x34..0x36].try_into().unwrap());
        let met_level = (meta & 0x7F) as u8;
        let game_of_origin = GameOfOriginIndex::from(((meta >> 7) & 0xF) as u8);
        let can_gigantamax = ((meta >> 11) & 1) != 0;
        let trainer_gender: Gender = (((meta >> 15) & 1) != 0).into();

        // IVs + flags (30 bits IVs starting at 0x36, then bits 30,31 flags)
        let ivs = Stats8::from_30_bits(bytes[0x36..0x3A].try_into().unwrap());
        let is_egg = util::get_flag(bytes, 0x36, 30);
        let has_hidden_ability = util::get_flag(bytes, 0x36, 31);

        // PP Ups (0x24): 2 bits per move
        let ppb = bytes[0x24];
        let move_pp_ups = [
            ppb & 0x3,
            (ppb >> 2) & 0x3,
            (ppb >> 4) & 0x3,
            (ppb >> 6) & 0x3,
        ];

        let mon = Pk3cfru {
            personality_value: u32::from_le_bytes(bytes[0..4].try_into().unwrap()),
            trainer_id: u16::from_le_bytes(bytes[4..6].try_into().unwrap()),
            secret_id: u16::from_le_bytes(bytes[6..8].try_into().unwrap()),
            nickname: Gen3String::from_bytes(bytes[8..18].try_into().unwrap()),
            language_index: bytes[18],
            trainer_name: Gen3String::from_bytes(bytes[19..26].try_into().unwrap()),
            markings: MarkingsFourShapes::from_byte(bytes[26]),
            species_and_forme: saf,
            cfru_species_index,
            held_item_index: u16::from_le_bytes(bytes[30..32].try_into().unwrap()),
            exp: u32::from_le_bytes(bytes[32..36].try_into().unwrap()),
            move_pp_ups,
            trainer_friendship: bytes[37],
            ball: cfru_ball_from_index(bytes[38]),
            moves: Self::read_moves_10bit(bytes),
            move_pp: [0, 0, 0, 0], // Not stored in CFRU; compute in UI layer if needed
            evs: Stats8::from_bytes(bytes[44..50].try_into().unwrap()),
            pokerus_byte: bytes[50],
            met_location_index: bytes[51],
            met_level,
            game_of_origin,
            can_gigantamax,
            trainer_gender,
            ivs,
            is_egg,
            has_hidden_ability,
            is_nicknamed: true, // TS default
            current_hp: 0,
        };

        Ok(mon)
    }
}

impl IsShiny for Pk3cfru {
    fn is_shiny(&self) -> bool {
        let tid = self.trainer_id as u32;
        let sid = self.secret_id as u32;
        let pid = self.personality_value;

        let xor = tid ^ sid ^ (pid & 0xFFFF) ^ ((pid >> 16) & 0xFFFF);
        xor < 8
    }
}

impl Pkm for Pk3cfru {
    const BOX_SIZE: usize = Pk3cfru::BOX_SIZE;
    const PARTY_SIZE: usize = Pk3cfru::PARTY_SIZE;

    fn box_size() -> usize {
        Self::BOX_SIZE
    }

    fn party_size() -> usize {
        Self::PARTY_SIZE
    }

    fn from_bytes(bytes: &[u8]) -> Result<Box<Self>> {
        Self::from_bytes(bytes).map(Box::new)
    }

    fn write_box_bytes(&self, bytes: &mut [u8]) {
        // Zero buffer then fill
        for b in bytes.iter_mut().take(Self::BOX_SIZE) {
            *b = 0;
        }

        // 0:4 Personality
        bytes[0..4].copy_from_slice(&self.personality_value.to_le_bytes());

        // 4:8 OT ID (Trainer ID and Secret ID)
        bytes[4..6].copy_from_slice(&self.trainer_id.to_le_bytes());
        bytes[6..8].copy_from_slice(&self.secret_id.to_le_bytes());

        // 8:18 Nickname (10 bytes)
        bytes[8..18].copy_from_slice(self.nickname.bytes().as_ref());

        // 18 Language
        bytes[18] = self.language_index;

        // 19 Sanity
        // bytes[19] = ???; // not yet implemented

        // 20:27 OT Name (7 bytes)
        bytes[20..27].copy_from_slice(self.trainer_name.bytes().as_ref());

        // 27 Markings
        bytes[27] = self.markings.to_byte();

        // 28:30 Species (DexNum / CFRU game index)
        if let Ok(cfru_species) =
            to_gen3_cfru_pokemon_index(&self.species_and_forme, &NATIONAL_DEX_TO_RADICAL_RED_MAP)
        {
            bytes[28..30].copy_from_slice(&cfru_species.to_le_bytes());
        } else {
            // fallback or propagate error
            bytes[28..30].copy_from_slice(&0u16.to_le_bytes());
        }

        // 30:32 Held Item
        bytes[30..32].copy_from_slice(&self.held_item_index.to_le_bytes());

        // 32:36 Experience
        bytes[32..36].copy_from_slice(&self.exp.to_le_bytes());

        // 36 PP Bonuses (PP Ups, 2 bits each)
        bytes[36] = (self.move_pp_ups[0] & 0x3)
            | ((self.move_pp_ups[1] & 0x3) << 2)
            | ((self.move_pp_ups[2] & 0x3) << 4)
            | ((self.move_pp_ups[3] & 0x3) << 6);

        // 37 Friendship
        bytes[37] = self.trainer_friendship;

        // 38 Ball (Poké Ball index)
        bytes[38] = cfru_ball_index(self.ball);

        // 39–43 Moves (4 × 10-bit packed into 5 bytes)
        Self::write_moves_10bit(&self.moves, bytes);

        // 44:50 EVs (6 stats, 1 byte each)
        bytes[44..50].copy_from_slice(&self.evs.to_bytes());

        // 50 Pokerus
        bytes[50] = self.pokerus_byte;

        // 51 Met Location
        bytes[51] = self.met_location_index;

        // 52:53 Met Info (packed: level, game of origin, Gigantamax, trainer gender)
        let mut meta: u16 = 0;
        meta |= (self.met_level as u16 & 0x7F) << 0;
        meta |= (u8::from(self.game_of_origin) as u16 & 0x0F) << 7;
        meta |= ((self.can_gigantamax as u16) & 0x01) << 11;
        meta |= ((bool::from(self.trainer_gender) as u16) & 0x01) << 15;
        bytes[52..54].copy_from_slice(&meta.to_le_bytes());

        // 54:58 IVs + Flags (30-bit IVs + 2 flag bits)
        self.ivs.write_30_bits(bytes, 54);
        util::set_flag(bytes, 54, 30, self.is_egg);
        util::set_flag(bytes, 54, 31, self.has_hidden_ability);
    }

    fn write_party_bytes(&self, bytes: &mut [u8]) {
        // CFRU uses the same 58 bytes; no extra party-only block provided.
        self.write_box_bytes(bytes);
    }

    fn to_box_bytes(&self) -> Vec<u8> {
        let mut bytes = vec![0; Self::BOX_SIZE];
        self.write_box_bytes(&mut bytes);
        bytes
    }

    fn to_party_bytes(&self) -> Vec<u8> {
        let mut bytes = vec![0; Self::PARTY_SIZE];
        self.write_party_bytes(&mut bytes);
        bytes
    }

    fn get_species_metadata(&self) -> &'static SpeciesMetadata {
        self.species_and_forme.get_species_metadata()
    }

    fn get_forme_metadata(&self) -> &'static FormeMetadata {
        self.species_and_forme.get_forme_metadata()
    }

    fn calculate_level(&self) -> u8 {
        self.get_species_metadata()
            .level_up_type
            .calculate_level(self.exp)
    }
}

File: src/core/save/cfru/PK3CFRU.ts
import { PluginPKMInterface, RomHackFormat } from '@openhome-core/pkm/interfaces'
import {
  Ball,
  ConvertStrategy,
  ExtraFormIndex,
  Language,
  Languages,
  MetadataLookup,
  NatureIndex,
  OriginGame,
  SpeciesLookup,
} from '@pkm-rs/pkg'
import { PkmConstructorOptions } from '@pokemon-files/pkm/PKM'
import {
  FourMoves,
  generatePersonalityValuePreservingAttributes,
  getFlag,
  getMoveMaxPP,
  getStandardPKMStats,
  MarkingsFourShapes,
  markingsFourShapesFromBytes,
  markingsFourShapesFromOther,
  markingsFourShapesToBytes,
  MoveFilter,
  read30BitIVsFromBytes,
  readGen3StringFromBytes,
  readStatsFromBytesU8,
  setFlag,
  Stats,
  uIntFromBufferBits,
  uIntToBufferBits,
  write30BitIVsToBytes,
  writeGen3StringToBytes,
  writeStatsToBytesU8,
} from '@pokemon-files/util'
import { PkmConverter } from '../../../../packages/pokemon-files/src/conversion/converter'
import { OHPKM } from '../../pkm/OHPKM'
import { Option } from '../../util/functional'
import { PluginIdentifier } from '../interfaces'
import { CfruSpeciesAndForm } from './conversion/util'

const INTERNAL_ORIGIN_NON_RR = OriginGame.Invalid6
const INTERNAL_ORIGIN_FROM_CFRU = OriginGame.FireRed
const FIRERED_IN_GAME_TRADE = 255

const CFRU_BALLS: Ball[] = [
  Ball.Master,
  Ball.Ultra,
  Ball.Great,
  Ball.Poke,
  Ball.Safari,
  Ball.Net,
  Ball.Dive,
  Ball.Nest,
  Ball.Repeat,
  Ball.Timer,
  Ball.Luxury,
  Ball.Premier,
  Ball.Dusk,
  Ball.Heal,
  Ball.Quick,
  Ball.Cherish,
  Ball.None,
  Ball.Fast,
  Ball.Level,
  Ball.Lure,
  Ball.Heavy,
  Ball.Love,
  Ball.Friend,
  Ball.Moon,
  Ball.PokeLegendsArceus,
  Ball.Beast,
  Ball.Dream,
]

export abstract class PK3CFRU implements PluginPKMInterface {
  // static getName() {
  //   return 'PK3RR'
  // }
  abstract format: RomHackFormat

  abstract pluginIdentifier: PluginIdentifier
  pluginOrigin?: PluginIdentifier
  personalityValue: number
  trainerID: number
  secretID: number
  language: Language
  markings: MarkingsFourShapes
  internalSpeciesIndex: number
  dexNum: number
  formeNum: number
  extraFormIndex: Option<ExtraFormIndex>
  internalHeldItemIndex: number
  abstract heldItemIndex: number
  exp: number
  movePPUps: FourMoves
  movePP: FourMoves = [0, 0, 0, 0]
  trainerFriendship: number
  moves: FourMoves
  evs: Stats
  pokerusByte: number
  internalMetLocationIndex: number
  metLocationIndex: number
  metLevel: number
  internalGameOfOrigin: number
  gameOfOrigin: number
  ball: number
  canGigantamax: boolean
  ivs: Stats
  isEgg: boolean
  hasHiddenAbility: boolean
  isNicknamed: boolean = true
  currentHP: number = 0
  nickname: string
  trainerName: string
  trainerGender: boolean
  isFakemon: boolean = false
  originalBytes?: Uint8Array

  pluginForm?: number

  abstract selectColor: string

  constructor(arg: ArrayBuffer | OHPKM, options: PkmConstructorOptions) {
    if (arg instanceof ArrayBuffer) {
      let buffer = arg
      const dataView = new DataView(buffer)

      this.originalBytes = new Uint8Array(arg)

      // https://github.com/Skeli789/Complete-Fire-Red-Upgrade/blob/master/include/new/pokemon_storage_system.h
      // https://github.com/Skeli789/Complete-Fire-Red-Upgrade/blob/master/include/pokemon.h

      // Personality 0:4
      this.personalityValue = dataView.getUint32(0x0, true)

      // OTID 4:8
      this.trainerID = dataView.getUint16(0x4, true)
      this.secretID = dataView.getUint16(0x6, true)

      // Nickname 8:18
      this.nickname = readGen3StringFromBytes(dataView, 0x8, 10)

      // Language 18
      this.language = Languages.fromByteOrNone(dataView.getUint8(0x12))

      // Sanity 19
      // this.sanity = dataView.getUint8(0x13)

      // OT Name 20:27
      this.trainerName = readGen3StringFromBytes(dataView, 0x14, 7)

      // Markings 27
      this.markings = markingsFourShapesFromBytes(dataView, 0x1b)

      // Species 28:30
      this.internalSpeciesIndex = dataView.getUint16(0x1c, true)
      const speciesData = this.monFromGameIndex(this.internalSpeciesIndex)
      this.extraFormIndex = speciesData.extraFormIndex

      if (speciesData.nationalDex < 0) {
        this.dexNum = 0
        this.formeNum = 0
      } else {
        this.dexNum = speciesData.nationalDex
        this.formeNum = speciesData.formIndex
      }

      this.isFakemon = this.indexIsFakemon(this.internalSpeciesIndex)

      // Held Item 30:32
      this.internalHeldItemIndex = dataView.getUint16(0x1e, true)

      // Exp 32:36
      this.exp = dataView.getUint32(0x20, true)

      // Move PP Up 36
      this.movePPUps = [
        uIntFromBufferBits(dataView, 0x24, 0, 2, true),
        uIntFromBufferBits(dataView, 0x24, 2, 2, true),
        uIntFromBufferBits(dataView, 0x24, 4, 2, true),
        uIntFromBufferBits(dataView, 0x24, 6, 2, true),
      ]

      // Friendship 37
      this.trainerFriendship = dataView.getUint8(0x25)

      // Pokeball 38
      const ballIndex = dataView.getUint8(0x26)

      this.ball = ballIndex < CFRU_BALLS.length ? CFRU_BALLS[ballIndex] : Ball.Poke

      // Moves 38:43 (5 bytes total for 4 moves with 10 bits each)
      this.moves = [
        this.moveFromGameIndex(uIntFromBufferBits(dataView, 0x27, 0, 10, true)), // Move 1
        this.moveFromGameIndex(uIntFromBufferBits(dataView, 0x28, 2, 10, true)), // Move 2
        this.moveFromGameIndex(uIntFromBufferBits(dataView, 0x29, 4, 10, true)), // Move 3
        this.moveFromGameIndex(uIntFromBufferBits(dataView, 0x2a, 6, 10, true)), // Move 4
      ]

      for (let i = 0; i < 4; i++) {
        const pp = getMoveMaxPP(this.moves[i], this.getFormat(), this.movePPUps[i])

        if (pp) this.movePP[i] = pp
      }

      // EVs 43:49
      this.evs = readStatsFromBytesU8(dataView, 0x2c)

      // 49
      this.pokerusByte = dataView.getUint8(0x32)

      // 50
      this.metLocationIndex = dataView.getUint8(0x33)
      this.internalMetLocationIndex = this.metLocationIndex

      // 51:53
      this.metLevel = uIntFromBufferBits(dataView, 0x34, 0, 7, true)

      this.gameOfOrigin = uIntFromBufferBits(dataView, 0x34, 7, 4, true)
      this.internalGameOfOrigin = this.gameOfOrigin

      // OHPKM will not copy this field if the mon was already being tracked before being
      // transferred to Radical Red
      this.pluginOrigin =
        this.gameOfOrigin === INTERNAL_ORIGIN_FROM_CFRU ? this.getPluginIdentifier() : undefined

      this.canGigantamax = getFlag(dataView, 0x34, 11)
      this.trainerGender = getFlag(dataView, 0x34, 15)

      // 53:57
      this.ivs = read30BitIVsFromBytes(dataView, 0x36)
      this.isEgg = getFlag(dataView, 0x36, 30)
      this.hasHiddenAbility = getFlag(dataView, 0x36, 31)
    } else {
      const other = arg
      const converter = new PkmConverter(this.getFormat(), options.strategy)
      const metData = converter.metData(other)

      this.personalityValue = generatePersonalityValuePreservingAttributes(other) ?? 0
      this.trainerID = other.trainerID
      this.secretID = other.secretID
      this.language = other.language
      this.markings = markingsFourShapesFromOther(other.markings)
      this.dexNum = other.dexNum
      this.formeNum = other.formeNum
      this.extraFormIndex = other.extraFormIndex
      this.internalHeldItemIndex = this.internalItemIndexFromModern(other.heldItemIndex)
      this.exp = other.exp
      this.trainerFriendship = other.trainerFriendship ?? 0

      const moveFilter = MoveFilter.fromMoveIndices(this.getValidMoveIndices())
      this.moves = moveFilter.moves(other)
      this.movePP = moveFilter.movePp(other, this.getFormat())
      this.movePPUps = moveFilter.movePpUps(other)

      this.evs = other.evs
      this.pokerusByte = other.pokerusByte
      this.metLocationIndex = metData.locationIndex
      this.metLevel = other.metLevel

      const fromRadicalRed = other.pluginOrigin === 'radical_red'

      if (fromRadicalRed) {
        this.gameOfOrigin = INTERNAL_ORIGIN_FROM_CFRU
        this.internalGameOfOrigin = this.gameOfOrigin

        this.metLocationIndex = other.metLocationIndex ?? FIRERED_IN_GAME_TRADE
        this.internalMetLocationIndex = this.metLocationIndex
      } else {
        this.gameOfOrigin = other.gameOfOrigin ?? INTERNAL_ORIGIN_NON_RR
        this.internalGameOfOrigin = INTERNAL_ORIGIN_NON_RR

        this.metLocationIndex = other.metLocationIndex ?? FIRERED_IN_GAME_TRADE
        this.internalMetLocationIndex = FIRERED_IN_GAME_TRADE
      }

      this.internalSpeciesIndex = this.monToGameIndex(other.dexNum, other.formeNum)

      if (other.pluginOrigin === this.getPluginIdentifier()) {
        this.pluginOrigin = other.pluginOrigin
      }

      if (other.ball) {
        this.ball =
          other.ball >= Ball.PokeLegendsArceus && other.ball <= Ball.Origin
            ? Ball.PokeLegendsArceus
            : other.ball === Ball.Sport
              ? Ball.Poke
              : other.ball
      } else {
        this.ball = Ball.Poke
      }

      this.canGigantamax = !!other.canGigantamax
      this.ivs = converter.ivs(other)
      this.isEgg = other.isEgg ?? false
      this.hasHiddenAbility = other.abilityNum === 4
      this.isNicknamed = other.isNicknamed ?? false
      this.currentHP = other.currentHP ?? 0
      this.nickname = other.nickname
      this.trainerName = other.trainerName
      this.trainerGender = other.trainerGender
    }
  }

  static fromBytes<T extends PK3CFRU>(
    this: new (buffer: ArrayBuffer, options: PkmConstructorOptions) => T,
    buffer: ArrayBuffer,
    encrypted?: boolean
  ): T {
    return new this(buffer, { encrypted })
  }

  static fromOhpkm<T extends PK3CFRU>(
    this: new (ohpkm: OHPKM, options: PkmConstructorOptions) => T,
    ohpkm: OHPKM,
    strategy: ConvertStrategy
  ): T {
    return new this(ohpkm, { strategy })
  }

  abstract internalItemIndexFromModern(modernIndex: number): number
  abstract itemToString(index: number): string

  abstract moveFromGameIndex(gameIndex: number): number
  abstract moveToGameIndex(nationalMoveId: number): number
  abstract getValidMoveIndices(): number[]

  abstract monFromGameIndex(gameIndex: number): CfruSpeciesAndForm
  abstract monToGameIndex(nationalDexNumber: number, formIndex: number): number

  abstract indexIsFakemon(speciesIndex: number): boolean

  toBytes(): ArrayBuffer {
    const buffer = new ArrayBuffer(58) // 58 bytes as specified
    const dataView = new DataView(buffer)

    // 0:4 Personality
    dataView.setUint32(0x0, this.personalityValue, true)

    // 4:8 OT ID (Trainer ID and Secret ID)
    dataView.setUint16(0x4, this.trainerID, true)
    dataView.setUint16(0x6, this.secretID, true)

    // 8:18 Nickname (10 bytes)
    writeGen3StringToBytes(dataView, this.nickname, 0x8, 10, false)

    // 18 Language
    dataView.setUint8(0x12, this.language)

    // 19 Sanity
    // dataView.setUint8(0x13, SANITY VALUE IDK);

    // 20:27 OT Name (7 bytes)
    writeGen3StringToBytes(dataView, this.trainerName, 0x14, 7, false)

    // 27 Markings
    markingsFourShapesToBytes(dataView, 0x1b, this.markings)

    // Growth Substructure (starts at 0x1C)
    // 28:30 Species (DexNum)
    if (this.pluginForm) {
      dataView.setUint16(0x1c, this.pluginForm, true)
    } else {
      dataView.setUint16(0x1c, this.monToGameIndex(this.dexNum, this.formeNum), true)
    }

    // 30:32 Held Item
    dataView.setUint16(0x1e, this.internalHeldItemIndex, true)

    // 32:36 Experience
    dataView.setUint32(0x20, this.exp, true)

    // 36 PP Bonuses (PP Ups)
    dataView.setUint8(
      0x24,
      this.movePPUps.reduce((acc, ppUp, i) => acc | (ppUp << (i * 2)), 0)
    )

    // 37 Friendship
    dataView.setUint8(0x25, this.trainerFriendship)

    // 38 Ball (Pokeball)
    const ballIndex =
      this.ball in CFRU_BALLS
        ? CFRU_BALLS.indexOf(this.ball)
        : this.ball >= Ball.PokeLegendsArceus && this.ball <= Ball.Origin
          ? Ball.PokeLegendsArceus
          : Ball.Poke

    dataView.setUint8(0x26, ballIndex)

    // Moves (5 bytes total for 10-bit moves)
    uIntToBufferBits(dataView, this.moveToGameIndex(this.moves[0]), 0x27, 0, 10, true)
    uIntToBufferBits(dataView, this.moveToGameIndex(this.moves[1]), 0x28, 2, 10, true)
    uIntToBufferBits(dataView, this.moveToGameIndex(this.moves[2]), 0x29, 4, 10, true)
    uIntToBufferBits(dataView, this.moveToGameIndex(this.moves[3]), 0x2a, 6, 10, true)

    // EVs
    writeStatsToBytesU8(dataView, 0x2c, this.evs)

    // 49 Pokerus
    dataView.setUint8(0x32, this.pokerusByte)

    // 50 Met Location
    dataView.setUint8(0x33, this.metLocationIndex)

    // 51:52 Met Info (packed: Met level, Game of Origin, Trainer Gender)
    uIntToBufferBits(dataView, this.metLevel, 0x34, 0, 7, true)
    uIntToBufferBits(dataView, this.internalGameOfOrigin, 0x34, 7, 4, true)
    setFlag(dataView, 0x34, 11, this.canGigantamax)
    setFlag(dataView, 0x34, 15, this.trainerGender)

    // 53:57 IVs and Flags (30-bit IVs + 2 bits for flags)
    write30BitIVsToBytes(dataView, 0x36, this.ivs)
    setFlag(dataView, 0x36, 30, this.isEgg)
    setFlag(dataView, 0x36, 31, this.hasHiddenAbility)

    return buffer
  }

  public getStats() {
    return getStandardPKMStats(this)
  }

  public get gender() {
    return this.metadata?.genderFromPid(this.personalityValue)
  }

  public get languageString() {
    return Languages.stringFromByte(this.language)
  }

  public get heldItemName() {
    return this.itemToString(this.internalHeldItemIndex)
  }

  public get nature() {
    return NatureIndex.newFromPid(this.personalityValue)
  }

  public get abilityNum() {
    return this.hasHiddenAbility ? 4 : ((this.personalityValue >> 0) & 1) + 1
  }

  public get ability() {
    return this.metadata?.abilityByNum(this.abilityNum)
  }

  toPCBytes() {
    return this.toBytes()
  }

  public getLevel() {
    return this.speciesMetadata?.calculateLevel(this.exp) ?? 1
  }

  isShiny() {
    return (
      (this.trainerID ^
        this.secretID ^
        (this.personalityValue & 0xffff) ^
        ((this.personalityValue >> 16) & 0xffff)) <
      8
    )
  }

  isSquareShiny() {
    return !(
      this.trainerID ^
      this.secretID ^
      (this.personalityValue & 0xffff) ^
      ((this.personalityValue >> 16) & 0xffff)
    )
  }

  public get metadata() {
    return MetadataLookup(this.dexNum, this.formeNum)
  }

  public get speciesMetadata() {
    return SpeciesLookup(this.dexNum)
  }

  static maxValidMove() {
    return 354
  }

  static maxValidBall() {
    return Ball.Beast
  }

  static allowedBalls() {
    return []
  }

  getPluginIdentifier(): PluginIdentifier {
    return this.pluginIdentifier
  }

  extraDisplayFields() {
    return {
      'Internal Species Index': this.internalSpeciesIndex,
    }
  }

  abstract getFormat(): RomHackFormat
}

export default PK3CFRU
